SICP_book/chapter_2_1.metta

```diff
@@ -0,0 +1,793 @@
+; -----------------------End of Exercise 2.7----------------------------
+
+; Exercise 2.8.
+;
+; Using reasoning analogous to Alyssa's, describe how the difference of two
+; intervals may be computed. Define a corresponding subtraction procedure, called sub-interval.
+
+(= (sub-interval $x $y)
+    (make-interval (- (lower-bound $x) (lower-bound $y))
+                   (- (upper-bound $x) (upper-bound $y))))
+
+!(assertEqual
+    (sub-interval (r2) (r1))
+    (0.75 . 0.9000000000000001))
+
+; -----------------------End of Exercise 2.8----------------------------
+
+; Exercise 2.9.
+;
+; The width of an interval is half of the difference between its upper and lower bounds.
+; The width is a measure of the uncertainty of the number specified by the interval.
+; For some arithmetic operations the width of the result of combining two intervals
+; is a function only of the widths of the argument intervals, whereas for others the width
+; of the combination is not a function of the widths of the argument intervals. Show that
+; the width of the sum (or difference) of two intervals is a function only of the widths
+; of the intervals being added (or subtracted). Give examples to show that this is not true
+; for multiplication or division.
+
+(= (interval-width $interval)
+    (/ (- (upper-bound $interval) (lower-bound $interval)) 2))
+
+!(assertEqual
+    (interval-width (r2))
+    0.10000000000000003)
+
+; Well, width of sum-intervals indeed equal to sum of widths, but due to rounding inside metta I can't directly compare
+; results using assertEqual. But we will consider 0.125 and 0.12500000000000006 equal.
+!(assertEqual
+    (interval-width (add-interval (r1) (r2)))
+    0.125)
+
+!(assertEqual
+    (+ (interval-width (r2)) (interval-width (r1)))
+    0.12500000000000006)
+
+; Same here
+
+!(assertEqual
+    (interval-width (sub-interval (r2) (r1)))
+    0.07500000000000007)
+
+!(assertEqual
+    (- (interval-width (r2)) (interval-width (r1)))
+    0.07500000000000002)
+
+; Different results for multiplication and division:
+!(assertEqual
+    (interval-width (mul-interval (r2) (r1)))
+    0.04250000000000001)
+
+!(assertEqual
+    (* (interval-width (r2)) (interval-width (r1)))
+    0.002500000000000002)
+
+!(assertEqual
+    (interval-width (div-interval (r2) (r1)))
+    1.416666666666667)
+
+!(assertEqual
+    (/ (interval-width (r2)) (interval-width (r1)))
+    4.0)
+; -----------------------End of Exercise 2.9----------------------------
+
+; Exercise 2.10.
+;
+; Ben Bitdiddle, an expert systems programmer, looks over Alyssa's shoulder and
+; comments that it is not clear what it means to divide by an interval that spans zero.
+; Modify Alyssa's code to check for this condition and to signal an error if it occurs.
+
+
+(= (better-div-interval $x $y)
+    (if (or (and (< (upper-bound $y) 0) (> (lower-bound $y) 0))
+             (and (< (lower-bound $y) 0) (> (upper-bound $y) 0)))
+        (Error: Divisor interval spans zero)
```

Comment:
I would use standard MeTTa error constructor here an below `(Error <atom> <message>)`. For example here: `(Error (better-div-interval $x $y) "Divisor interval spans zero")`.

Reply:
Will do.